### PR TITLE
Add support for IsLocal in DAG.getAddr()

### PR DESCRIPTION
**Pull request by arichardson**

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVExpandPseudoInsts.cpp [hunk 1]
```diff
@@ -83,6 +83,9 @@ class RISCVExpandPseudo : public MachineFunctionPass {
   bool expandCapLoadGlobalCap(MachineBasicBlock &MBB,
                               MachineBasicBlock::iterator MBBI,
                               MachineBasicBlock::iterator &NextMBBI);
+  bool expandCapLoadLocalCap(MachineBasicBlock &MBB,
```
Comment (jrtc27) [inline llvm/lib/Target/RISCV/RISCVExpandPseudoInsts.cpp]: This should come before Global.

File: llvm/lib/Target/RISCV/RISCVExpandPseudoInsts.cpp [hunk 2]
```diff
@@ -1166,6 +1171,13 @@ bool RISCVExpandPseudo::expandCapLoadGlobalCap(
                               SecondOpcode);
 }
 
+bool RISCVExpandPseudo::expandCapLoadLocalCap(
```
Comment (jrtc27) [inline llvm/lib/Target/RISCV/RISCVExpandPseudoInsts.cpp]: Before Global again.

**Comment by jrtc27** [comment]
This doesn't work for dynamic linking; `shouldAssumeDSOLocal` will return true AFAIK, and will need a PLT stub, which isn't yet implemented. Not hard to do, but needs to happen first.

**Comment by arichardson** [comment]
We could use IsLocal=false for calls, that way at least constant pools avoid the captable.

**Comment by arichardson** [comment]
Update without changing calls.

Turns out if we change the generated code for calls the following reduced test case fails to compile with the following error in RISCVAsmBackend:
```
 case RISCV::fixup_riscv_branch: {
    if (!isInt<13>(Value))
      Ctx.reportError(Fixup.getLoc(), "fixup value out of range");
```
```
; RUN: llc -o /dev/null -mtriple=riscv64-unknown-freebsd13.0 -relocation-model=pic -thread-model=posix -mattr=+m -mattr=+a -mattr=+f -mattr=+d -mattr=+c -mattr=+xcheri -mattr=+cap-mode -mattr=-relax -target-abi l64pc128d -vectorize-loops -vectorize-slp -O0 -verify-machineinstrs -filetype=obj %s
; ModuleID = '/Users/alex/cheri/build/llvm-project-build/large-branch-reduce.ll-reduced-simplified.bc'
source_filename = "/Users/alex/cheri/llvm-project/llvm/test/CodeGen/RISCV/cheri/large-branch-reduce.ll"
target datalayout = "e-m:e-pf200:128:128:128:64-p:64:64-i64:64-i128:128-n64-S128-A200-P200-G200"
target datalayout = "e-m:e-pf200:128:128:128:64-p:64:64-i64:64-i128:128-n64-S128-A200-P200-G200"
target triple = "x86_64-apple-darwin18.7.0"

%struct.ham = type {}
%struct.blam = type {}

@global = external addrspace(200) constant [12 x i8]
@global.1 = external addrspace(200) constant [27 x i8]
@global.2 = external addrspace(200) constant [5 x i8]
@global.3 = external addrspace(200) constant [24 x i8]
@global.4 = external addrspace(200) constant [5 x i8]
@global.5 = external addrspace(200) constant [4 x i8]
declare i32 @widget(%struct.ham addrspace(200)*, i8 addrspace(200)*, ...) addrspace(200)
declare i64 @bar(i8 addrspace(200)*) addrspace(200)
declare i8 addrspace(200)* @bar.6(i8 addrspace(200)*, i64) addrspace(200)
declare i8 @blam(%struct.blam addrspace(200)*, i8 addrspace(200)*) addrspace(200)
declare i16 @eggs(%struct.blam addrspace(200)*, i8 addrspace(200)*) addrspace(200)
declare hidden i32 @barney(%struct.blam addrspace(200)*, i8 addrspace(200)*) addrspace(200)
declare hidden i8 addrspace(200)* @pluto(%struct.blam addrspace(200)*, i8 addrspace(200)*) addrspace(200)
define i8 @wibble(i32 %arg) addrspace(200) {
bb:
  switch i32 %arg, label %bb95 [
    i32 3, label %bb30
    i32 6, label %bb35
    i32 8, label %bb37
    i32 0, label %bb40
    i32 2, label %bb44
    i32 4, label %bb47
    i32 15, label %bb49
    i32 20, label %bb54
    i32 21, label %bb57
    i32 26, label %bb63
    i32 27, label %bb65
    i32 30, label %bb69
    i32 143, label %bb72
    i32 146, label %bb93
    i32 38, label %bb74
    i32 39, label %bb76
    i32 47, label %bb78
    i32 40, label %bb80
    i32 42, label %bb87
    i32 46, label %bb89
    i32 48, label %bb90
    i32 101, label %bb91
    i32 113, label %bb92
    i32 121, label %bb92
    i32 145, label %bb94
  ]

bb30:                                             ; preds = %bb
  %tmp = call i8 addrspace(200)* @pluto(%struct.blam addrspace(200)* undef, i8 addrspace(200)* undef)
  %tmp31 = call i8 addrspace(200)* @bar.6(i8 addrspace(200)* null, i64 16)
  %tmp32 = call i64 @bar(i8 addrspace(200)* undef)
  %tmp34 = call i32 (%struct.ham addrspace(200)*, i8 addrspace(200)*, ...) @widget(%struct.ham addrspace(200)* undef, i8 addrspace(200)* getelementptr inbounds ([12 x i8], [12 x i8] addrspace(200)* @global, i64 0, i64 0))
  unreachable

bb35:                                             ; preds = %bb
  %tmp36 = call i8 addrspace(200)* @pluto(%struct.blam addrspace(200)* undef, i8 addrspace(200)* undef)
  unreachable

bb37:                                             ; preds = %bb
  %tmp38 = call i8 @blam(%struct.blam addrspace(200)* null, i8 addrspace(200)* undef)
  %tmp39 = call i64 @bar(i8 addrspace(200)* null)
  unreachable

bb40:                                             ; preds = %bb
  %tmp41 = call i8 addrspace(200)* @bar.6(i8 addrspace(200)* undef, i64 undef)
  %tmp42 = getelementptr inbounds i8 addrspace(200)*, i8 addrspace(200)* addrspace(200)* undef, i32 1
  store i8 addrspace(200)* addrspace(200)* %tmp42, i8 addrspace(200)* addrspace(200)** null, align 16
  store i8 addrspace(200)* null, i8 addrspace(200)* addrspace(200)* %tmp42
  %tmp43 = call i64 @bar(i8 addrspace(200)* null)
  unreachable

bb44:                                             ; preds = %bb
  %tmp45 = call signext i32 @barney(%struct.blam addrspace(200)* null, i8 addrspace(200)* undef)
  %tmp46 = call i8 addrspace(200)* @bar.6(i8 addrspace(200)* undef, i64 undef)
  unreachable

bb47:                                             ; preds = %bb
  %tmp48 = call i64 @bar(i8 addrspace(200)* undef)
  unreachable

bb49:                                             ; preds = %bb
  %tmp50 = call i8 addrspace(200)* @bar.6(i8 addrspace(200)* null, i64 8)
  %tmp51 = call i64 @bar(i8 addrspace(200)* null)
  %tmp52 = call i64 @bar(i8 addrspace(200)* %tmp50)
  %tmp53 = call i8 addrspace(200)* @bar.6(i8 addrspace(200)* null, i64 undef)
  unreachable

bb54:                                             ; preds = %bb
  %tmp55 = load i8 addrspace(200)*, i8 addrspace(200)* addrspace(200)* undef, align 16
  store i8 addrspace(200)* %tmp55, i8 addrspace(200)** null
  %tmp56 = call signext i32 (%struct.ham addrspace(200)*, i8 addrspace(200)*, ...) @widget(%struct.ham addrspace(200)* undef, i8 addrspace(200)* getelementptr inbounds ([27 x i8], [27 x i8] addrspace(200)* @global.1, i64 0, i64 0))
  unreachable

bb57:                                             ; preds = %bb
  %tmp58 = call zeroext i8 @blam(%struct.blam addrspace(200)* undef, i8 addrspace(200)* null)
  %tmp59 = call i8 addrspace(200)* @bar.6(i8 addrspace(200)* null, i64 1)
  %tmp60 = call i64 @bar(i8 addrspace(200)* undef)
  %tmp61 = call i64 @bar(i8 addrspace(200)* %tmp59)
  %tmp62 = call i8 addrspace(200)* @bar.6(i8 addrspace(200)* undef, i64 undef)
  unreachable

bb63:                                             ; preds = %bb
  %tmp64 = call i8 addrspace(200)* @bar.6(i8 addrspace(200)* undef, i64 undef)
  unreachable

bb65:                                             ; preds = %bb
  %tmp66 = load i8 addrspace(200)* addrspace(200)*, i8 addrspace(200)* addrspace(200)* addrspace(200)* null
  %tmp67 = load i8 addrspace(200)*, i8 addrspace(200)* addrspace(200)* %tmp66, align 16
  store i8 addrspace(200)* %tmp67, i8 addrspace(200)** undef
  %tmp68 = call signext i32 (%struct.ham addrspace(200)*, i8 addrspace(200)*, ...) @widget(%struct.ham addrspace(200)* undef, i8 addrspace(200)* getelementptr inbounds ([5 x i8], [5 x i8] addrspace(200)* @global.2, i64 0, i64 0))
  unreachable

bb69:                                             ; preds = %bb
  %tmp70 = call i64 @bar(i8 addrspace(200)* null)
  %tmp71 = call i64 @bar(i8 addrspace(200)* undef)
  unreachable

bb72:                                             ; preds = %bb
  %tmp73 = call i8 addrspace(200)* @bar.6(i8 addrspace(200)* null, i64 undef)
  unreachable

bb74:                                             ; preds = %bb
  %tmp75 = call signext i32 (%struct.ham addrspace(200)*, i8 addrspace(200)*, ...) @widget(%struct.ham addrspace(200)* undef, i8 addrspace(200)* getelementptr inbounds ([24 x i8], [24 x i8] addrspace(200)* @global.3, i64 0, i64 0))
  unreachable

bb76:                                             ; preds = %bb
  %tmp77 = call signext i32 (%struct.ham addrspace(200)*, i8 addrspace(200)*, ...) @widget(%struct.ham addrspace(200)* undef, i8 addrspace(200)* getelementptr inbounds ([5 x i8], [5 x i8] addrspace(200)* @global.4, i64 0, i64 0))
  unreachable

bb78:                                             ; preds = %bb
  %tmp79 = call i8 addrspace(200)* @bar.6(i8 addrspace(200)* undef, i64 undef)
  unreachable

bb80:                                             ; preds = %bb
  %tmp81 = call zeroext i16 @eggs(%struct.blam addrspace(200)* null, i8 addrspace(200)* undef)
  %tmp82 = call i8 addrspace(200)* @bar.6(i8 addrspace(200)* null, i64 2)
  %tmp83 = load i8 addrspace(200)*, i8 addrspace(200)** undef
  %tmp84 = call i64 @bar(i8 addrspace(200)* %tmp82)
  %tmp85 = call i8 addrspace(200)* @bar.6(i8 addrspace(200)* %tmp83, i64 undef)
  %tmp86 = call i8 addrspace(200)* @bar.6(i8 addrspace(200)* null, i64 undef)
  unreachable

bb87:                                             ; preds = %bb
  tail call void asm sideeffect ".space 1000", ""()
  %tmp88 = call signext i32 (%struct.ham addrspace(200)*, i8 addrspace(200)*, ...) @widget(%struct.ham addrspace(200)* undef, i8 addrspace(200)* getelementptr inbounds ([4 x i8], [4 x i8] addrspace(200)* @global.5, i64 0, i64 0))
  unreachable

bb89:                                             ; preds = %bb
  tail call void asm sideeffect ".space 1000", ""()
  unreachable

bb90:                                             ; preds = %bb
  tail call void asm sideeffect ".space 1000", ""()
  unreachable

bb91:                                             ; preds = %bb
  unreachable

bb92:                                             ; preds = %bb, %bb
  unreachable

bb93:                                             ; preds = %bb
  unreachable

bb94:                                             ; preds = %bb
  unreachable

bb95:                                             ; preds = %bb
  unreachable
}

```

**Comment by jrtc27** [comment]
Oh you’re trying to use this for globals as well? That’s really not a good idea; permissions will be RX^W, but you will also have no bounds for your accesses, i.e. you’ve just completely undone the benefit of having purecap? You can only do this for read-only accesses, and you can only do this when you can statically prove all accesses are in bounds (and don’t violate load cap permissions either). You’re also going to introduce copy relocations that we don’t currently have.

**Comment by arichardson** [comment]
Well I wanted it for RO globals, but you're right this does it for all globals.

**Pull request closed without merging**